### PR TITLE
Containers: support using OCI base images.

### PR DESCRIPTION
**Pull request by jeffpardy**
Adds support for using oci base images from #35609

### CODE REVIEW

**Comment by jeffpardy** [comment]
Thanks @nagilson! @baronfel if possible would benefit from having this backported into 8.0.2xx and 8.0.3xx after it lands in main.

**Comment by baronfel** [comment]
This looks reasonable to me - how can we test it, though? Do you know of any images/hosts that are OCI-only? Our current Accepts headers to me suggest that we're preferencing the Docker manifest format over the OCI one, so the only way we'd ever see this is if a Registry only supported OCI manifests.

**Comment by baronfel** [comment]
We can easily do a backport to 8.0.300, but 8.0.200 would probably not meet the bar for servicing without some critical customer impact.

**Comment by jeffpardy** [comment]
Sure happy to share our issue, similar to original report in #35608. We are currently experiencing this using [JFrog Artifactory](https://jfrog.com/artifactory/) as repository. We build custom runtime images on top of `mcr.microsoft.com/dotnet/[runtime|aspnet]` using `podman/buildah`

This results in a base image manifest like this:
```
{
        "schemaVersion": 2,
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "config": {
                "mediaType": "application/vnd.oci.image.config.v1+json",
        },
        "layers": [
                {
                        "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
                },
               // additional layers
        ]
}
```
Then when using that base image with containers sdk we get:
```
{
        "schemaVersion": 2,
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "config": {
                "mediaType": "application/vnd.oci.image.config.v1+json",
        },
        "layers": [
                {
                        "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
                },
                {
                        // layer from container sdk
                        "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
                }
        ]
}
```

Finally when trying to push that image to ECR (again using `podman`) we get this error
```
unsupported MIME type for compression: application/vnd.docker.image.rootfs.diff.tar.gzip
```
This might eventually be addressed with https://github.com/containers/podman/issues/19949, but seems like we should build layers matching either OCI or Docker manifest types to avoid validation on mixed layer types.

> We can easily do a backport to 8.0.300, but 8.0.200 would probably not meet the bar for servicing without some critical customer impact.

I think 8.0.300 would be fine 👍🏼

**Comment by baronfel** [comment]
Yep, that makes sense to me. Thanks for the explanation.

**Review by baronfel — APPROVED** [review]

File: src/Containers/Microsoft.NET.Build.Containers/Layer.cs [hunk 1]
```diff
@@ -186,9 +187,17 @@ static UnixFileMode DetermineFileMode(FileSystemInfo file)
         string contentHash = Convert.ToHexString(hash).ToLowerInvariant();
         string uncompressedContentHash = Convert.ToHexString(uncompressedHash).ToLowerInvariant();
 
+        string layerMediaType = manifestMediaType switch
+        {
+             // TODO: configurable? gzip always?
+            SchemaTypes.DockerManifestV2 => "application/vnd.docker.image.rootfs.diff.tar.gzip",
+            SchemaTypes.OciManifestV1 => "application/vnd.oci.image.layer.v1.tar+gzip",
```
Comment (baronfel) [inline src/Containers/Microsoft.NET.Build.Containers/Layer.cs:194]: Would you mind extracting these to constants like we've slowly been doing for the other media types?

**Pull request merged**